### PR TITLE
feat: withdeletiontimestamp filter

### LINKED ISSUE #1018: `.WithDeletionTimestamp` Filter
### Is your feature request related to a problem? Please describe.
UDS-Core is using finalizers to control deletion of managed objects. 

They were using `When(Object).isDeleted(..)` but this no longer works because the finalizer simply `Updates` instead of deletes because it adds a `deletionTimestamp` to the object metadata.

Definition of done:
- [ ] unit test
- [ ] e2e test
- [ ] works for mutate, validate, watch, reconcile

### Describe the solution you'd like

- **Given** someone is using Pepr to write an Operator and the operator uses finalizers
- **When** an update occurs
- **Then** they can use the `WithDeletionTimestamp` filter to check if they need to "operate" on the object

### Describe alternatives you've considered
(optional) A clear and concise description of any alternative solutions or features you've considered.

### LINKED ISSUE #1027: Adds docs around the Pepr filters
### Is your feature request related to a problem? Please describe.
The filters available in Pepr are not clearly documented in the docs.

### Describe the solution you'd like

- **Given** someone is reading the docs and wants to know what filters are available 
- **When** they look at the filters section 
- **Then** they know

### Describe alternatives you've considered
(optional) A clear and concise description of any alternative solutions or features you've considered.

### PR DESCRIPTION
**Pull request by cmwylie19**
## Description

Adds a new filter `.WithDeletionTimestamp` to aid Operator developers as they use `ownerRefs`. Decide if `.HasDeletionTimestamp` needs to be done. 

Needs:
- [x] e2e test (https://github.com/defenseunicorns/pepr-excellent-examples/pull/61)
- [x] docs 

## Related Issue

Fixes #1018 
Fixes: #1027 

Relates to #

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://docs.pepr.dev/main/contribute/#submitting-a-pull-request) followed

### CODE REVIEW

**Comment by cmwylie19** [comment]
> ## [Codecov](https://app.codecov.io/gh/defenseunicorns/pepr/pull/1026?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=defenseunicorns) Report
> Attention: Patch coverage is `64.28571%` with `5 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 85.55%. Comparing base [(`5c0b2d9`)](https://app.codecov.io/gh/defenseunicorns/pepr/commit/5c0b2d9fb63e586a9e0b7c4bac7dabc74734fa7f?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=defenseunicorns) to head [(`073ea49`)](https://app.codecov.io/gh/defenseunicorns/pepr/commit/073ea49d56e90ecbb07ff6fe28d1d6b8da754eb6?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=defenseunicorns).
> 
> [Files](https://app.codecov.io/gh/defenseunicorns/pepr/pull/1026?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=defenseunicorns)	Patch %	Lines
> [src/lib/capability.ts](https://app.codecov.io/gh/defenseunicorns/pepr/pull/1026?src=pr&el=tree&filepath=src%2Flib%2Fcapability.ts&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=defenseunicorns#diff-c3JjL2xpYi9jYXBhYmlsaXR5LnRz)	0.00%	[5 Missing ⚠️ ](https://app.codecov.io/gh/defenseunicorns/pepr/pull/1026?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=defenseunicorns)
> Additional details and impacted files

false positive, this is tested in other ways and does not drop the overall coverage.

File: src/lib/types.ts [hunk 1]
```diff
@@ -137,8 +138,14 @@ export type BindingFilter<T extends GenericClass> = CommonActionChain<T> & {
...
    */
   WithAnnotation: (key: string, value?: string) => BindingFilter<T>;
+  /** Only apply the action if the resource has a deletionTimestamp. */
+  WithDeletionTimestamp: () => BindingFilter<T>;
 };
 
+// export type BindingWithDeletionTimestamp<T extends GenericClass> = BindingFilter<T> & {
+
+// };
+
```
Comment (cmwylie19) [inline src/lib/types.ts]:
```suggestion
```

File: src/lib/helpers.ts [hunk 2]
```diff
@@ -73,6 +73,11 @@ export function filterNoMatchReason(
   obj: Partial<KubernetesObject>,
   capabilityNamespaces: string[],
 ): string {
+  // binding deletionTimestamp filter and object deletionTimestamp dont match
+  if (binding.filters && binding.filters.deletionTimestamp && obj.metadata && !obj.metadata.deletionTimestamp) {
```
Comment (samayer12) [inline src/lib/helpers.ts]: Could this be `binding.filters?.deletionTimestamp`? It'd just be one less conditional (`&&`) to think about while still achieving the same result, I think.

Comment (samayer12) [reply src/lib/helpers.ts]: Similarly, I think `!obj.metadata?.deletionTimestamp` gets the same result with one less set of `&&`. But I'll defer on if the syntax to express _"Not 'If `metadata` exists then is `deletionTimestamp` truthy?'"_ is easier to read.

File: src/lib/helpers.test.ts [hunk 3]
```diff
@@ -1056,6 +1056,40 @@ describe("filterMatcher", () => {
...
+    };
+    const capabilityNamespaces: string[] = [];
+    const result = filterNoMatchReason(
+      binding as unknown as Partial<Binding>,
+      obj as unknown as Partial<KubernetesObject>,
+      capabilityNamespaces,
+    );
+    expect(result).not.toEqual("Ignoring Watch Callback: Object does not have a deletion timestamp.");
+  });
+
```
Comment (samayer12) [inline src/lib/helpers.test.ts:1092]: Woah, long file! Not an issue here, but I wonder if there are ways to reorganize/de-duplicate stuff in a file this long

File: src/lib/filter.ts [hunk 4]
```diff
@@ -22,6 +22,20 @@ export function shouldSkipRequest(binding: Binding, req: AdmissionRequest, capab
...
 
+  // Delete bindings do not work through admission with DeletionTimestamp
+  if (binding.event.includes(Event.Delete) && binding.filters?.deletionTimestamp) {
+    return true;
+  }
+
+  // Test for deletionTimestamp
+  if (
+    binding.filters &&
+    binding.filters.deletionTimestamp &&
```
Comment (samayer12) [inline src/lib/filter.ts]: Could this be `binding.filters?.deletionTimestamp`? It'd just be one less conditional (`&&`) to think about while still achieving the same result, I think.

File: src/lib/filter.ts [hunk 5]
```diff
@@ -22,6 +22,20 @@ export function shouldSkipRequest(binding: Binding, req: AdmissionRequest, capab
...
+    req.object.metadata &&
+    !req.object.metadata?.deletionTimestamp
```
Comment (samayer12) [inline src/lib/filter.ts]: Seeing a code-style trend for checking `X.prop` and then drilling into `X.prop.subprop` that might also be common in Pepr or provides some benefit I'm unaware of compared to `?` for optional-chaining

**Review by samayer12 — COMMENTED** [review]
Aside from a stylistic choice on multiple `&&` statements vs. Optional Chaining with`?` things look good. I'll still have to get more familiar with our tests

File: src/lib/helpers.test.ts (hunk 3 above)
Comment (cmwylie19) [reply src/lib/helpers.test.ts:1092]: yeah theoretically we could pull out that long CapabilityExport object and have it in a separate file

File: src/lib/filter.ts (hunk 4 above)
Comment (cmwylie19) [reply src/lib/filter.ts]: resolved in PR

File: docs/030_user-guide/130_filters.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+## `Filters`
+
+- `.WithName("name")`: Filters resources by name.
+- `.InNamespace("namespace")`: Filters resources by namespace.
+- `.WithLabel("key", "value")`: Filters resources by label. (Can be multiple)
+- `.WithDeletionTimestamp()`: Filters resources that have a deletion timestamp. 
+
+Notes: 
+- `WithDeletionTimestamp()` is does not work on Delete through the `Mutate` or `Validate` methods because the Kubernetes Admission Process does not fire the DELETE event with a deletion timestamp on the resource. 
+- `WithDeletionTimestamp()` is fired on an Update event during Admission (`Mutate` or `Validate`) _only_ when the finalizer is removed from the resource.
```
Comment (btlghrants) [inline docs/030_user-guide/130_filters.md]:
```suggestion
- `WithDeletionTimestamp()` _will_ match on an Update event during Admission (`Mutate` or `Validate`) when pending-deletion permitted changes (like removing a finalizer) occur.
```

**Review by btlghrants — APPROVED** [review]
Seems legit -- glad we chatted it through.

**Comment by cmwylie19** [comment]
CodeCov is false positive, the coverage remains at 85.

**Pull request merged**